apollo-ios/Sources/Apollo/FieldPolicyEvaluator.swift

```diff
@@ -0,0 +1,190 @@
+import Foundation
+import ApolloAPI
+
+enum FieldPolicyResult {
+  case single(CacheKeyInfo)
+  case list([CacheKeyInfo])
+}
+
+struct FieldPolicyEvaluator {
+  let field: Selection.Field
+  let variables: GraphQLOperation.Variables?
+  
+  init(
+    field: Selection.Field,
+    variables: GraphQLOperation.Variables?
+  ) {
+    self.field = field
+    self.variables = variables
+  }
+  
+  func resolveFieldPolicy() -> FieldPolicyResult? {
+    guard let fieldPolicy = field.fieldPolicy,
+          let arguments = field.arguments else {
+      return nil
+    }
+    
+    struct ParsedKey { let name: String; let path: [String]? }
+
+    let parsed: [ParsedKey] = fieldPolicy.keys.map { key in
+      if let dot = key.firstIndex(of: ".") {
+        let name = String(key[..<dot])
+        let rest = key[key.index(after: dot)...]
+        return ParsedKey(name: name, path: rest.split(separator: ".").map(String.init))
+      } else {
+        return ParsedKey(name: key, path: nil)
+      }
+    }
+
+    var fixedParts = [String?](repeating: nil, count: parsed.count)
+    var listIndex: Int? = nil
+    var listValues: [String] = []
+
+    for (i, pk) in parsed.enumerated() {
+      guard let argVal = arguments[pk.name] else {
+        return nil
+      }
+
+      guard let resolved = argVal.resolveValue(keyPath: pk.path, variables: variables), !resolved.isEmpty else {
+        return nil
+      }
+      
+      if resolved.count > 1 {
+        listIndex = i
+        listValues = resolved
+      } else {
+        guard let value = resolved.first else {
+          return nil
+        }
+        fixedParts[i] = value
+      }
+    }
+
+    if let idx = listIndex {
+      var keys: [CacheKeyInfo] = []
+      keys.reserveCapacity(listValues.count)
+      for item in listValues {
+        var parts = [String]()
+        parts.reserveCapacity(parsed.count)
+        for j in 0..<parsed.count {
+          if j == idx {
+            parts.append(item)
+          } else if let v = fixedParts[j] {
+            parts.append(v)
+          } else {
+            return nil
+          }
+        }
+        keys.append(CacheKeyInfo(id: parts.joined(separator: "+")))
+      }
+      return .list(keys)
+    } else {
+      let parts = fixedParts.compactMap { $0 }
+      guard parts.count == parsed.count else {
+        return nil
```

Comment:
Is there a scenario in which this actually should happen and return `nil`? Or is this actually an error that isn't being thrown?